jwst/datamodels/util.py

```diff
@@ -488,3 +493,157 @@ def get_envar_as_boolean(name, default=False):
+        size *= dimension
+
+    # Get available memory
+    available = get_available_memory(include_swap=include_swap)
+    log.debug(f'Model size {bytes2human(size)} available system memory {bytes2human(available)}')
+
+    if size > available:
+        log.warning(
+            f'Model {model_type} shape {shape} requires {bytes2human(size)} which is more than'
+            f' system available {bytes2human(available)}'
+        )
+
+    if allowed and size > (allowed * available):
+        log.debug(
+            f'Model size greater than allowed memory {bytes2human(allowed * available)}'
+        )
+        return False, size
+
+    return True, size
+
+
+def get_available_memory(include_swap=True):
+    """Retrieve available memory
+
+    Parameters
+    ----------
+    include_swap : bool
+        Include available swap in the calculation.
+
+    Returns
+    -------
+    available : number
+        The amount available.
+    """
+    system = platform_system()
+
+    # Apple MacOS
+    log.debug(f'Running OS is "{system}"')
+    if system in ['Darwin']:
+        return get_available_memory_darwin(include_swap=include_swap)
+
+    # Default to Linux-like:
+    return get_available_memory_linux(include_swap=include_swap)
+
+
+def get_available_memory_linux(include_swap=True):
+    """Get memory for a Linux system
+
+    Presume that the swap space as reported is accurate at the time of
+    the query and that any subsequent allocation will be held the value.
+
+    Parameters
+    ----------
+    include_swap : bool
+        Include available swap in the calculation.
+
+    Returns
+    -------
+    available : number
+        The amount available.
+    """
+    vm_stats = psutil.virtual_memory()
+    available = vm_stats.available
+    if include_swap:
+        swap = psutil.swap_memory()
+        available += swap.total
+    return available
+
+
+def get_available_memory_darwin(include_swap=True):
+    """Get the available memory on an Apple MacOS-like system
+
+    For Darwin, swap space is dynamic and will attempt to use the whole of the
+    boot partition.
+
+    If the system has been configured to use swap from other sources besides
+    the boot partition, that available space will not be included.
+
+    Parameters
+    ----------
+    include_swap : bool
+        Include available swap in the calculation.
+
+    Returns
+    -------
+    available : number
+        The amount available.
+    """
+    vm_stats = psutil.virtual_memory()
+    available = vm_stats.available
+    if include_swap:
+
+        # Attempt to determine amount of free disk space on the boot partition.
+        run_info = run(['/usr/sbin/bless', '--getBoot', '--plist'], capture_output=True)
+        boot_volume = plist_loads(run_info.stdout)['Boot Volume']
+        partitions = psutil.disk_partitions()
+        for partition in partitions:
+            if partition.device == boot_volume:
+                swap = psutil.disk_usage(partition.mountpoint).free
```

Comment:
Maybe leave 10% or so of the boot volume for temp files or whatever else it needs?

Reply:
With the default being "no check", and otherwise the user having "free reign" on how they want to manage their memory, I have a slight preference for not having any hidden padding in there. But not wedded to either way.

Reply:
That makes sense, fine by me.